calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,380 @@
+// or done.
+fn if_name_stable_or_done(assign: &ir::Assignment, name: &ir::Id) -> bool {
+    let reads = ReadWriteSet::port_reads(assign);
+    reads
+        .filter(|port_ref| port_ref.borrow().get_parent_name() == name)
+        .all(|port_ref| {
+            let atts = &port_ref.borrow().attributes;
+            atts.has("stable") || atts.has("done")
+        })
+}
+
+// Returns true if the cell is a component or a non-combinational primitive
+fn comp_or_non_comb(cell: &ir::RRC<ir::Cell>) -> bool {
+    match &cell.borrow().prototype {
+        ir::CellType::Primitive { is_comb, .. } => !*is_comb,
+        ir::CellType::Component { .. } => true,
+        _ => false,
+    }
+}
+
+//If asmt is a write to a cell named name returns Some(name).
+//If asmt is a write to a group port, returns None.
+fn writes_to_cell(asmt: &ir::Assignment) -> Option<ir::Id> {
+    match &asmt.dst.borrow().parent {
+        ir::PortParent::Cell(cell) => {
+            Some(cell.upgrade().borrow().clone_name())
+        }
+        ir::PortParent::Group(_) => None,
+    }
+}
+
+#[derive(Default)]
+///Primarily used to help determine the order cells are executed within
+///the group, and if possible, to transform a group into a seq of two smaller groups
+pub struct SplitAnalysis {
+    // First 2 fields help determine if transformation can be applied. Remaining
+    // fields help to apply the transformation.
+    ///Holds, (b,a) for assignment of form a.go = b.done,
+    done_go: Option<(ir::Id, ir::Id)>,
+
+    ///Holds a for assignment of form group[done] = a.done
+    last: Option<ir::Id>,
+
+    /// Holds the go-done assignment, i.e. a.go = b.done
+    go_done_asmt: Option<ir::Assignment>,
+
+    /// Holds the first "go" assignment, *if* it is in the form a.go = !a.done ? 1'd1
+    first_go_asmt: Option<ir::Assignment>,
+
+    /// Holds the group[done] = done assignment;
+    group_done_asmt: Option<ir::Assignment>,
+
+    /// Assignments that write to first cell, unless the assignment is already accounted by a different field
+    fst_asmts: Vec<ir::Assignment>,
+
+    /// Assignments that write to second cell, unless the assignment is already accounted by a different field
+    snd_asmts: Vec<ir::Assignment>,
+}
+
+impl SplitAnalysis {
+    /// Based on assigns, returns Ok(group1, group2), where (group1,group2) are
+    /// the groups that can be made by splitting assigns. If it is not possible to split
+    /// assigns into two groups, then just regurn Err(assigns).
+    /// Criteria for being able to split assigns into two groups:
+    /// 1) Group must write to exactly 2 cells -- let's call them cell1 and cell2
+    /// 2) cell1 and cell2 must be either non-combinational primitives or components
+    /// 3) Must have group[done] = cell2.done and cell2.go = cell1.done;
+    /// 4) All reads of cell1 must be a stable port or cell1.done.
+    pub fn get_split(
+        assigns: &[ir::Assignment],
```

Comment:
This should take ownership of the assigns instead of taking a slice. The `to_vec` call below is a clone operation which should be avoided